app/lib/actions/katello/repository/sync.rb

```diff
@@ -124,6 +126,62 @@ def regenerate_applicability(repo, contents_changed)
         def rescue_strategy
           Dynflow::Action::Rescue::Skip
         end
+
+        def repository_id
+          input['repository']['id']
+        end
+
+        def repository_name
+          input['repository']['name']
+        end
+
+        def repository_label
+          input['repository']['label']
+        end
+
+        def product_id
+          input['product']['id']
+        end
+
+        def product_name
+          input['product']['name']
+        end
+
+        def product_label
+          input['product']['label']
+        end
+
+        def contents_changed
+          input['contents_changed']
+        end
+
+        def sync_result
+          input['sync_result']
+        end
+        apipie :class, "A class representing #{self} object" do
+          desc 'This object is available as **@object** variable in
+                webhook templates when a corresponding event occures.
+                The following properties can be used to retrieve the needed information.'
+          name "#{class_scope}"
+          refs "#{class_scope}"
+          sections only: %w[all webhooks]
+          property :task, object_of: 'Task', desc: 'Returns the task to which this action belongs'
+          property :repository_id, Integer, desc: 'Returns synced repository id'
+          property :repository_name, String, desc: 'Returns synced repository name'
+          property :repository_label, String, desc: 'Returns synced repository label'
+          property :product_id, Integer, desc: 'Returns product id the synced repository belongs to'
+          property :product_name, String, desc: 'Returns product name the synced repository belongs to'
+          property :product_label, String, desc: 'Returns product label the synced repository belongs to'
+          property :sync_result, Hash, desc: 'Returns Hash object with sync result'
+          property :contents_changed, one_of: [true, false], desc: 'Returns true if repository content was changed due to sync, false otherwise'
+        end
+        include Actions::Katello::JailConcern::Organization
+        class Jail < ::Actions::ObservableAction::Jail
+          allow :organization_id, :organization_name, :organization_label,
+                :repository_id, :repository_name, :repository_label,
+                :product_id, :product_name, :product_label,
+                :sync_result, :contents_changed
+        end
```

Comment:
Do I understand right that the `apipie` definition, the include JailConcern and the class Jail declaration are all tied to the same concept? Events?

Reply:
Yes, for `actions` these are needed to fully support events in Foreman Webhooks that are being fired up if e.g. repo sync has happened.

`apipie` block is for documentation; JailConcern with public methods to use in webhook templates; Jails are to make those methods available in safe mode.

Reply:
Thanks. It seems like it would be useful to group all of the pieces related to webhooks together somehow but I'll let Katello core maintainers weigh in on the overall code structure. I appreciate the helpful details understanding. I was worried that it would be hard over time to track what is required to ensure Webhooks work (and subsequently do not break in the future) and what aspects are unrelated. Also, how to help developers who want to add additinal webhook "hooks" into the code know what to add.